src/python/pants/backend/docker/rules_test.py

```diff
@@ -0,0 +1,256 @@
+    )
+
+
+@pytest.mark.parametrize(
+    "target_values, expected_features",
+    [
+        (
+            dict(),
+            dict(
+                context_root="docker/test/.",
+            ),
+        ),
+        (
+            dict(
+                version="1.2.3",
+            ),
+            dict(
+                version="1.2.3",
+            ),
+        ),
+        (
+            dict(
+                build_root="/",
+            ),
+            dict(
+                context_root=".",
+            ),
+        ),
+        (
+            dict(
+                build_root="foo/bar",
+            ),
+            dict(
+                context_root="docker/test/foo/bar",
+            ),
+        ),
+        (
+            dict(
+                build_root="/foo/bar",
+            ),
+            dict(
+                context_root="foo/bar",
+            ),
+        ),
+    ],
+)
+def test_build_docker_image(target_values, expected_features):
+    address = Address("docker/test", target_name="image")
+    image = DockerImage(
+        address=address,
+        unhydrated_values=target_values,
+    )
+    targets = {address: TransitiveTargets(roots=(image,), dependencies=FrozenOrderedSet())}
+    field_set = DockerFieldSet.create(image)
+
+    def build_context_mock(request: DockerBuildContextRequest) -> DockerBuildContext:
+        if "context_root" in expected_features:
+            assert expected_features["context_root"] == request.context_root
+
+        return DockerBuildContext(digest=EMPTY_DIGEST)
+
+    result = run_rule_with_mocks(
+        build_docker_image,
+        rule_args=[field_set],
+        mock_gets=[
+            MockGet(
+                output_type=DockerBinary,
+                input_type=DockerBinaryRequest,
+                mock=lambda _: DockerBinary("/dummy/docker"),
+            ),
+            MockGet(
+                output_type=TransitiveTargets,
+                input_type=TransitiveTargetsRequest,
+                mock=lambda request: targets[request.roots[0]],
+            ),
+            MockGet(
+                output_type=DockerBuildContext,
+                input_type=DockerBuildContextRequest,
+                mock=build_context_mock,
+            ),
+            MockGet(
+                output_type=Snapshot,
+                input_type=Digest,
+                mock=lambda _: EMPTY_SNAPSHOT,
+            ),
+            MockGet(
+                output_type=ProcessResult,
+                input_type=Process,
+                # Process() generation has its own tests in test_docker_build_image
+                mock=lambda _: ProcessResult(
+                    stdout=b"stdout",
+                    stdout_digest=EMPTY_FILE_DIGEST,
+                    stderr=b"stderr",
+                    stderr_digest=EMPTY_FILE_DIGEST,
+                    output_digest=EMPTY_DIGEST,
+                ),
+            ),
+        ],
+        # union_membership=
```

Comment:
?

Reply:
wasn't sure if it was needed when I copied from another test. will clean up.